### PR TITLE
feat(elasticsearch): Add servlet which provides an endpoint for a healthcheck on the ES cl…

### PR DESCRIPTION
**Pull request by iprentic**
…ient

## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: metadata-service/health-servlet/src/main/java/com/datahub/health/controller/HealthCheckController.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+import org.springframework.web.bind.annotation.RestController;
+
+
+@RestController
+@AllArgsConstructor
+@RequestMapping("/check")
+@Tag(name = "HealthCheck", description = "An API for checking health of GMS and its clients.")
+public class HealthCheckController {
+  @Autowired
+  @Qualifier("elasticSearchRestHighLevelClient")
```
[the remark this answers is not in the document]
Comment (pedro93) [reply metadata-service/health-servlet/src/main/java/com/datahub/health/controller/HealthCheckController.java:29]: This warning seems relevant. 
We have an all args constructor for all fields in this class but one of them is injected by spring. Could this be a cause for concern?

File: metadata-service/health-servlet/src/main/java/com/datahub/health/controller/HealthCheckController.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+      responseString = response.toString();
+      if (isHealthy) {
+        return ResponseEntity.ok(responseString);
+      }
+    } catch (Exception e) {
+      if (responseString == null) {
+        // Couldn't get a response, fill in the string from the exception message
+        responseString = e.getMessage();
+      }
+    } finally {
```
[the remark this answers is not in the document]
Comment (pedro93) [reply metadata-service/health-servlet/src/main/java/com/datahub/health/controller/HealthCheckController.java:67]: Question: Is there case where the code catches an exception and the response string is not null?
In that case we will not log or return the exception as exemplified in code scan alert details.

Suggestion: Can we remove this finally block and return the response outside the try catch block?

File: metadata-service/health-servlet/src/main/java/com/datahub/health/controller/HealthCheckController.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+  public HealthCheckController () {
+    this.memoizedSupplier = Suppliers.memoizeWithExpiration(
+        this::getElasticHealth, 5, TimeUnit.SECONDS);
+  }
+
+  /**
+   * Checks the memoized cache for the latest elastic health check result
+   * @return The ResponseEntity containing the health check result
+   */
+  @GetMapping(path = "/elastic", produces = MediaType.APPLICATION_JSON_VALUE)
```
Comment (david-leifker) [inline metadata-service/health-servlet/src/main/java/com/datahub/health/controller/HealthCheckController.java:42]: Where is the combined healthcheck endpoint. Specifically, the one that produces a common message with all of the component checks? The one which would be used by a readiness check.

File: metadata-service/health-servlet/src/main/java/com/datahub/health/controller/HealthCheckController.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+@RequestMapping("/check")
+@Tag(name = "HealthCheck", description = "An API for checking health of GMS and its clients.")
+public class HealthCheckController {
+  @Autowired
+  @Qualifier("elasticSearchRestHighLevelClient")
+  private RestHighLevelClient elasticClient;
+  private Supplier<ResponseEntity<String>> memoizedSupplier;
+
+  public HealthCheckController () {
+    this.memoizedSupplier = Suppliers.memoizeWithExpiration(
```
Comment (david-leifker) [inline metadata-service/health-servlet/src/main/java/com/datahub/health/controller/HealthCheckController.java:34]: Nice!

File: metadata-service/health-servlet/src/main/java/com/datahub/health/controller/HealthCheckController.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (david-leifker) [reply metadata-service/health-servlet/src/main/java/com/datahub/health/controller/HealthCheckController.java:67]: return the response outside the finally, inside the finally set the responseString to something if it is null or empty

File: metadata-service/health-servlet/src/main/java/com/datahub/health/controller/HealthCheckController.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (RyanHolstien) [reply metadata-service/health-servlet/src/main/java/com/datahub/health/controller/HealthCheckController.java:28]: You can get around this and still do Constructor injection by specifying the qualifier in the Constructor explicitly. This removes the need for the Autowired annotation. That being said we kind of mix our usage of these so it's fine that it doesn't use Constructor injection although using the Autowired annotation is not generally the recommended way to inject stuff :)

**Review by david-leifker — APPROVED** [review]

**Pull request merged**